### PR TITLE
Refactor ffi again

### PR DESCRIPTION
**Pull request by jim-portegies**
Move the ffi to its own file, so there is a better separation of functionality, and so developers can use the ocaml lsp server for feedback.

Also remove the obsolete `Framework.v` file.

### CODE REVIEW

File: src/wp_ffi.ml [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+(** Creates a name used to define the function interface *)
+let pname (s: string): ml_tactic_name = { mltac_plugin = "coq-core.plugins.coq-waterproof"; mltac_tactic = s }
+
+(** Wrapper around {! Tac2env.define_primitive} to make easier the primitive definition *)
+let define_primitive (name: string) (arity: 'a arity) (f: 'a): unit =
+  Tac2env.define_primitive (pname name) (mk_closure arity f)
+
+(** 
+  Defines a function of arity 0 (that only take a [unit] as an argument)
```
Comment (pimotte) [inline src/wp_ffi.ml:44]:
```suggestion
  Defines a function of arity 0 (that only takes a [unit] as an argument)
```

File: src/wp_ffi.ml [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+  This function will be callable in Ltac2 with [Ltac2 @ external <ltac2_name>: unit := "coq-waterproof" "<name>".]
+*)
+let define0 (name: string) (f: valexpr tactic): unit = define_primitive name arity_one (fun _ -> f)
+
+(** 
+  Defines a function of arity 1 (that only take one argument)
```
Comment (pimotte) [inline src/wp_ffi.ml:51]:
```suggestion
  Defines a function of arity 1 (that only takes one argument)
```

File: src/wp_ffi.ml [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+  This function will be callable in Ltac2 with [Ltac2 @ external <ltac2_name>: <type> -> unit := "coq-waterproof" "<name>".]
+*)
+let define1 (name: string) (r0: 'a repr) (f: 'a -> valexpr tactic): unit =
+  define_primitive name arity_one @@ fun x -> f (repr_to r0 x)
+
+(** 
+  Defines a function of arity 2 of the same way than {! define1}
```
Comment (pimotte) [inline src/wp_ffi.ml:59]:
```suggestion
  Defines a function of arity 2 in the same way as {! define1}
```

File: src/wp_ffi.ml [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+*)
+let define2 (name: string) (r0: 'a repr) (r1: 'b repr) (f: 'a -> 'b -> valexpr tactic): unit =
+  define_primitive name (arity_suc arity_one) @@ fun x y -> f (repr_to r0 x) (repr_to r1 y)
+
+(** 
+  Defines a function of arity 3 of the same way than {! define1}
```
Comment (pimotte) [inline src/wp_ffi.ml:65]:
```suggestion
  Defines a function of arity 3 in the same way as {! define1}
```

File: src/wp_ffi.ml [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+*)
+let define3 (name: string) (r0: 'a repr) (r1: 'b repr) (r2: 'c repr) (f: 'a -> 'b -> 'c -> valexpr tactic): unit =
+  define_primitive name (arity_suc (arity_suc arity_one)) @@ fun x y z -> f (repr_to r0 x) (repr_to r1 y) (repr_to r2 z)
+
+(** 
+  Defines a function of arity 4 of the same way than {! define1}
```
Comment (pimotte) [inline src/wp_ffi.ml:71]:
```suggestion
  Defines a function of arity 4 in the same way as {! define1}
```

File: src/wp_ffi.ml [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+*)
+let define4 (name: string) (r0: 'a repr) (r1: 'b repr) (r2: 'c repr) (r3: 'd repr) (f: 'a -> 'b -> 'c -> 'd -> valexpr tactic): unit =
+  define_primitive name (arity_suc (arity_suc (arity_suc arity_one))) @@
+  fun x0 x1 x2 x3 -> f (repr_to r0 x0) (repr_to r1 x1) (repr_to r2 x2) (repr_to r3 x3)
+
+(** 
+  Defines a function of arity 5 of the same way than {! define1}
```
Comment (pimotte) [inline src/wp_ffi.ml:78]:
```suggestion
  Defines a function of arity 5 in the same way as {! define1}
```

File: src/wp_ffi.ml [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+*)
+let define5 (name: string) (r0: 'a repr) (r1: 'b repr) (r2: 'c repr) (r3: 'd repr) (r4: 'e repr) (f: 'a -> 'b -> 'c -> 'd -> 'e -> valexpr tactic): unit =
+  define_primitive name (arity_suc (arity_suc (arity_suc (arity_suc arity_one)))) @@
+  fun x0 x1 x2 x3 x4 -> f (repr_to r0 x0) (repr_to r1 x1) (repr_to r2 x2) (repr_to r3 x3) (repr_to r4 x4)
+
+(** 
+  Defines a function of arity 6 of the same way than {! define1}
```
Comment (pimotte) [inline src/wp_ffi.ml:85]:
```suggestion
  Defines a function of arity 6 in the same way as {! define1}
```

File: src/wp_ffi.ml [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+*)
+let define6 (name: string) (r0: 'a repr) (r1: 'b repr) (r2: 'c repr) (r3: 'd repr) (r4: 'e repr) (r5: 'f repr) (f: 'a -> 'b -> 'c -> 'd -> 'e -> 'f -> valexpr tactic): unit =
+  define_primitive name (arity_suc (arity_suc (arity_suc (arity_suc (arity_suc arity_one))))) @@
+  fun x0 x1 x2 x3 x4 x5 -> f (repr_to r0 x0) (repr_to r1 x1) (repr_to r2 x2) (repr_to r3 x3) (repr_to r4 x4) (repr_to r5 x5)
+
+(** 
+  Defines a function of arity 7 of the same way than {! define1}
```
Comment (pimotte) [inline src/wp_ffi.ml:92]:
```suggestion
  Defines a function of arity 7 in the same way as {! define1}
```

File: src/wp_ffi.ml [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+let delayed_of_tactic (tac: 'a tactic) (env: Environ.env) (sigma: Evd.evar_map): (Evd.evar_map * 'a) =
+  let _, pv = Proofview.init sigma [] in
+  let name, poly = Names.Id.of_string "ltac2_delayed", false in
+  let c, pv, _, _ = Proofview.apply ~name ~poly env tac pv in
+  let _, sigma = Proofview.proofview pv in
+  (sigma, c)
+
+
+(**
+  Utilitary function to cast OCaml types into Ltac2-compatibles types
```
Comment (pimotte) [inline src/wp_ffi.ml:111]:
```suggestion
  Utility function to cast OCaml types into Ltac2-compatibles types  
```

**Review by pimotte — APPROVED** [review]
Some small documentation-nitpicks, other than that this looks good to me, as far as my understanding goes:)

**Comment by jim-portegies** [comment]
Thanks for looking at it!

**Pull request merged**